testing/dsp_testing/test_beamforming.py

```diff
@@ -1,123 +1,693 @@
+                elif data_description_list == ['num_sequences', 'num_antennas', 'num_samps']:
+                    if filetype == tx_filetype:
+                        decimated_data = record_dict['data'][0,:,:] # first sequence only, all antennas.
+                    else:
+                        decimated_data = record_dict['data'][sequence_num,:,:] # first sequence only, all antennas.
+                else:
+                    raise Exception('Unexpected data dimensions: {}'.format(record_dict['data_descriptors']))
+
+            # STEP 2: BEAMFORM ANY UNBEAMFORMED DATA
+            if filetype != bfiq_filetype:
+                # need to beamform the data. 
+                antenna_list = []
+                #print(decimated_data.shape)
+                if data_description_list == ['num_antennas', 'num_sequences', 'num_samps']:
+                    for antenna in range(0, record_dict['data'].shape[0]):
+                        antenna_list.append(decimated_data[antenna,:])
+                    antenna_list = np.array(antenna_list)
+                elif data_description_list == ['num_sequences', 'num_antennas', 'num_samps']:
+                    for antenna in range(0, record_dict['data'].shape[1]):
+                        antenna_list.append(decimated_data[antenna,:])
+                    antenna_list = np.array(antenna_list)
+                else:
+                    raise Exception('Not sure how to beamform with the dimensions of this data: {}'.format(record_dict['data_descriptors']))
+
+                # beamform main array antennas only. 
+                main_antennas_mask = (record_dict['antennas_present'] < main_antenna_count)
+                intf_antennas_mask = (record_dict['antennas_present'] >= main_antenna_count)
+                decimated_beamformed_data = beamform(antenna_list[main_antennas_mask][:].copy(), beam_azms, freq) 
+                intf_decimated_beamformed_data = beamform(antenna_list[intf_antennas_mask][:].copy(), beam_azms, freq) 
+            else:
+                decimated_beamformed_data = decimated_data  
+                intf_decimated_beamformed_data = intf_decimated_data
+
+            sequence_filetype_dict['main_bf_data'] = decimated_beamformed_data # this has 2 dimensions: num_beams x num_samps for this sequence.
+            sequence_filetype_dict['intf_bf_data'] = intf_decimated_beamformed_data
+
+
+            # STEP 3: FIND THE PULSES IN THE DATA
+            for beamnum in range(0, sequence_filetype_dict['main_bf_data'].shape[0]):
+
+                len_of_data = sequence_filetype_dict['main_bf_data'].shape[1]
+                pulse_indices = find_pulse_indices(sequence_filetype_dict['main_bf_data'][beamnum], 0.09)
+                if len(pulse_indices) > len(pulses): # sometimes we get two samples from the same pulse.
+                    if math.fmod(len(pulse_indices), len(pulses)) == 0.0:
```

Comment:
this looks like it assumes that if one pulse gives you x number of indices from the find_pulse_indices(...) call, then all the pulses would give you the same amount of indices, is that the case in reality?

Reply:
except for those that missed. In our case mpinc is a multiple of sample rate so yes that should be the case.